File: packages/cli/src/core/redaction.ts
export type RedactionMode = "secret-redaction";

export interface RedactionFinding {
  pattern: string;
  replacements: number;
}

export interface RedactionResult {
  text: string;
  findings: RedactionFinding[];
  replacements: number;
}

interface RedactionPattern {
  id: string;
  regex: RegExp;
  replace: string | ((match: string, ...groups: string[]) => string);
}

const REDACTION_PATTERNS: RedactionPattern[] = [
  {
    id: "private_key",
    regex:
      /-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----/g,
    replace: "[REDACTED:private_key]",
  },
  {
    id: "github_token",
    regex: /\bgh[pousr]_[A-Za-z0-9_]{20,}\b/g,
    replace: "[REDACTED:github_token]",
  },
  {
    id: "npm_token",
    regex: /\bnpm_[A-Za-z0-9]{20,}\b/g,
    replace: "[REDACTED:npm_token]",
  },
  {
    id: "bearer_token",
    regex: /\bBearer\s+([A-Za-z0-9._~+/=-]{10,})\b/g,
    replace: "Bearer [REDACTED:bearer_token]",
  },
  {
    id: "jwt",
    regex: /\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b/g,
    replace: "[REDACTED:jwt]",
  },
  {
    id: "connection_string",
    regex: /\b((?:postgres(?:ql)?|mysql|mongodb|redis):\/\/)[^\s"'<>]+/gi,
    replace: (_match, protocol: string) =>
      `${protocol}[REDACTED:connection_string]`,
  },
  {
    id: "api_key",
    regex:
      /\b(api[_-]?key|apikey|access[_-]?key|secret[_-]?key)\s*[:=]\s*["']?([A-Za-z0-9._~+/=-]{12,})["']?/gi,
    replace: (_match, key: string) => `${key}=[REDACTED:api_key]`,
  },
  {
    id: "password_assignment",
    regex: /\b(password|passwd|pwd)\s*[:=]\s*["']?([^\s"'`]{6,})["']?/gi,
    replace: (_match, key: string) => `${key}=[REDACTED:password_assignment]`,
  },
];

export function redactText(
  input: string,
  _mode: RedactionMode = "secret-redaction"
): RedactionResult {
  let text = input;
  const findings: RedactionFinding[] = [];

  for (const pattern of REDACTION_PATTERNS) {
    let replacements = 0;
    text = text.replace(pattern.regex, (...args: string[]) => {
      replacements += 1;
      if (typeof pattern.replace === "function") {
        return pattern.replace(args[0], ...args.slice(1));
      }
      return pattern.replace;
    });
    if (replacements > 0) {
      findings.push({ pattern: pattern.id, replacements });
    }
  }

  return {
    text,
    findings,
    replacements: findings.reduce(
      (sum, finding) => sum + finding.replacements,
      0
    ),
  };
}

export function redactObject<T>(input: T): {
  value: T;
  result: RedactionResult;
} {
  const raw = JSON.stringify(input);
  const result = redactText(raw);
  return {
    value: JSON.parse(result.text) as T,
    result,
  };
}

File: internal/evidence/redaction.go
package evidence

import (
	"regexp"
	"strings"
)

type redactionPattern struct {
	id      string
	regex   *regexp.Regexp
	replace func(match string, groups []string) string
}

var redactionPatterns = []redactionPattern{
	{
		id: "private_key",
		regex: regexp.MustCompile(`-----BEGIN [A-Z ]*PRIVATE KEY-----[\s\S]*?-----END [A-Z ]*PRIVATE KEY-----`),
		replace: func(string, []string) string { return "[REDACTED:private_key]" },
	},
	{
		id: "github_token",
		regex: regexp.MustCompile(`\bgh[pousr]_[A-Za-z0-9_]{20,}\b`),
		replace: func(string, []string) string { return "[REDACTED:github_token]" },
	},
	{
		id: "npm_token",
		regex: regexp.MustCompile(`\bnpm_[A-Za-z0-9]{20,}\b`),
		replace: func(string, []string) string { return "[REDACTED:npm_token]" },
	},
	{
		id: "bearer_token",
		regex: regexp.MustCompile(`\bBearer\s+([A-Za-z0-9._~+/=-]{10,})\b`),
		replace: func(match string, groups []string) string {
			return "Bearer [REDACTED:bearer_token]"
		},
	},
	{
		id: "jwt",
		regex: regexp.MustCompile(`\beyJ[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\.[A-Za-z0-9_-]+\b`),
		replace: func(string, []string) string { return "[REDACTED:jwt]" },
	},
	{
		id: "connection_string",
		regex: regexp.MustCompile(`(?i)\b((?:postgres(?:ql)?|mysql|mongodb|redis):\/\/)[^\s"'<>]+`),
		replace: func(match string, groups []string) string {
			if len(groups) > 0 {
				return groups[0] + "[REDACTED:connection_string]"
			}
			return "[REDACTED:connection_string]"
		},
	},
	{
		id: "api_key",
		regex: regexp.MustCompile(`(?i)\b(api[_-]?key|apikey|access[_-]?key|secret[_-]?key)\s*[:=]\s*["']?([A-Za-z0-9._~+/=-]{12,})["']?`),
		replace: func(match string, groups []string) string {
			if len(groups) > 0 {
				return groups[0] + "=[REDACTED:api_key]"
			}
			return "[REDACTED:api_key]"
		},
	},
	{
		id: "password_assignment",
		regex: regexp.MustCompile("(?i)\\b(password|passwd|pwd)\\s*[:=]\\s*[\"']?([^\\s\"'`]{6,})[\"']?"),
		replace: func(match string, groups []string) string {
			if len(groups) > 0 {
				return groups[0] + "=[REDACTED:password_assignment]"
			}
			return "[REDACTED:password_assignment]"
		},
	},
}

// RedactText applies redaction patterns to text content.
// It returns the redacted content, matched pattern IDs, and total replacements.
func RedactText(content string) (redacted string, patterns []string, replacements int) {
	text := content
	var matchedPatterns []string
	totalReplacements := 0

	for _, pattern := range redactionPatterns {
		count := 0
		text = pattern.regex.ReplaceAllStringFunc(text, func(match string) string {
			groups := pattern.regex.FindStringSubmatch(match)
			var subgroups []string
			if len(groups) > 1 {
				subgroups = groups[1:]
			}
			count++
			return pattern.replace(match, subgroups)
		})
		if count > 0 {
			matchedPatterns = append(matchedPatterns, pattern.id)
			totalReplacements += count
		}
	}

	return text, matchedPatterns, totalReplacements
}

// IsTextFile returns true if the file extension indicates a text file,
// or if the content contains no null bytes.
func IsTextFile(path string, content []byte) bool {
	ext := strings.ToLower(path)
	for _, suffix := range textExtensions {
		if strings.HasSuffix(ext, suffix) {
			return true
		}
	}
	return !bytesContainsNull(content)
}

func bytesContainsNull(b []byte) bool {
	for _, v := range b {
		if v == 0 {
			return true
		}
	}
	return false
}

var textExtensions = []string{
	".txt", ".md", ".json", ".jsonl", ".yaml", ".yml",
	".log", ".out", ".err", ".stdout", ".stderr", ".env",
	".ts", ".js", ".tsx", ".jsx", ".go", ".sh", ".mod", ".sum",
}
